chains/substrate/client/client.go

```diff
@@ -0,0 +1,141 @@
+// The Licensed Work is (c) 2022 Sygma
+// SPDX-License-Identifier: BUSL-1.1
+
+package client
+
+import (
+	"fmt"
+	"math/big"
+	"sync"
+
+	"github.com/centrifuge/go-substrate-rpc-client/v4/types/codec"
+
+	"github.com/ChainSafe/sygma-relayer/chains/substrate/connection"
+	"github.com/centrifuge/go-substrate-rpc-client/v4/signature"
+	"github.com/centrifuge/go-substrate-rpc-client/v4/types"
+	"github.com/rs/zerolog/log"
+)
+
+type SubstrateClient struct {
+	key       *signature.KeyringPair // Keyring used for signing
+	nonceLock sync.Mutex             // Locks nonce for updates
+	nonce     types.U32              // Latest account nonce
+
+	Conn    *connection.Connection
+	ChainID *big.Int
+}
+
+func NewSubstrateClient(conn *connection.Connection, key *signature.KeyringPair, chainID *big.Int) *SubstrateClient {
+	return &SubstrateClient{
+		key:     key,
+		Conn:    conn,
+		ChainID: chainID,
+	}
+}
+
+// Transact constructs and submits an extrinsic to call the method with the given arguments.
+// All args are passed directly into GSRPC. GSRPC types are recommended to avoid serialization inconsistencies.
+func (c *SubstrateClient) Transact(method string, args ...interface{}) (*types.Hash, error) {
+	log.Debug().Msgf("Submitting substrate call... method %s, sender %s", method, c.key.Address)
+
+	// Create call and extrinsic
+	meta := c.Conn.GetMetadata()
+	call, err := types.NewCall(
+		&meta,
+		method,
+		args...,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("failed to construct call: %w", err)
+	}
+
+	ext := types.NewExtrinsic(call)
+	// Get latest runtime version
+	rv, err := c.Conn.RPC.State.GetRuntimeVersionLatest()
+	if err != nil {
+		return nil, err
+	}
+
+	c.nonceLock.Lock()
+	defer c.nonceLock.Unlock()
+
+	nonce, err := c.nextNonce(&meta)
+	if err != nil {
+		return nil, err
+	}
+
+	// Sign the extrinsic
+	o := types.SignatureOptions{
+		BlockHash:          c.Conn.GenesisHash,
+		Era:                types.ExtrinsicEra{IsMortalEra: false},
+		GenesisHash:        c.Conn.GenesisHash,
+		Nonce:              types.NewUCompactFromUInt(uint64(nonce)),
+		SpecVersion:        rv.SpecVersion,
+		Tip:                types.NewUCompactFromUInt(0),
+		TransactionVersion: rv.TransactionVersion,
+	}
+	h, err := c.signAndSendTransaction(o, ext)
```

Comment:
ext is passed without pointer but being copied first insdie signAndSendTransaction and then into sendRawTransaction. both times it allocates some heap memory bcs I beleive ext is a pretty big struct from the first glance. Although it is not critical now, we should kinda thinking about this to improve performance